### PR TITLE
Add exclude to initializer applicator, clean up types

### PR DESCRIPTION
**Pull request by matt-gardner**

### CODE REVIEW

File: allennlp/nn/initializers.py [hunk 1]
```diff
@@ -78,26 +118,33 @@ def __call__(self, module: torch.nn.Module) -> None:
...
-        initializers: {
-            parameter_regex_match1: {
-                "type": "normal"
-                "mean": 0.01
-                "std": 0.1
-            },
-            parameter_regex_match2: "uniform",
+        {
+            "initializers": {
+                "[parameter_regex_match1]": {
```
Comment (DeNeutoy) [inline allennlp/nn/initializers.py]: Why are there brackets here? It doesn't look like the functionality has changed.

File: allennlp/nn/initializers.py [hunk 2]
```diff
@@ -39,20 +72,27 @@ class InitializerApplicator:
...
         ----------
-        initializers : Dict[str, Callable[[torch.Tensor], None]], optional (default = {})
+        initializers : ``Dict[str, Callable[[torch.Tensor], None]]``, optional (default = {})
             A dictionary mapping parameter regexes to initializers to be applied to parameters
             matching the regex.
-        default_initializer : Callable[[torch.Tensor], None], optional, (default = torch.nn.init.normal)
+        default_initializer : ``Callable[[torch.Tensor], None]``, optional (default = torch.nn.init.normal)
             A default initializer, which will be used in the case that the Applicator encounters a parameter
             which does not match any of the regexes provided.
+        exclude : ``Sequence[str]``, optional (default=``[]``)
```
Comment (DeNeutoy) [inline allennlp/nn/initializers.py]: Maybe we should raise an error in the case that any exclude regex also matches an explicitly provided regex. Something's probably gone wrong if that case occurs. Alternatively the exclude regexes should override everything - doesn't seem like having this conditional application really buys us anything.

File: allennlp/nn/initializers.py (hunk 1 above)
Comment (matt-gardner) [inline allennlp/nn/initializers.py]: The brackets are intended to show that this isn't a literal string, it's something to be replaced with an actual regex.  I can remove them if you want.

File: allennlp/nn/initializers.py (hunk 2 above)
Comment (matt-gardner) [inline allennlp/nn/initializers.py]: The question for me is what's easier to understand and use?  If you have competing regexes around which things get initialized with what, it just seemed too confusing.  It's a lot simpler if you just say that exclusion only applies to things that would have been initialized with the default.

File: allennlp/nn/initializers.py (hunk 1 above)
Comment (DeNeutoy) [inline allennlp/nn/initializers.py]: Yes, they are confusing.

File: allennlp/nn/initializers.py (hunk 2 above)
Comment (DeNeutoy) [inline allennlp/nn/initializers.py]: sure.

**Pull request merged**